engine/src/test/java/io/camunda/zeebe/engine/util/EngineRule.java

```diff
@@ -423,28 +427,43 @@ private static class ProcessingExporterTransistor implements StreamProcessorLife
 
     private LogStreamReader logStreamReader;
     private TypedRecordImpl typedEvent;
+    private final SynchronousLogStream synchronousLogStream;
+    private final ExecutorService executorService;
+
+    public ProcessingExporterTransistor(final SynchronousLogStream synchronousLogStream) {
+      this.synchronousLogStream = synchronousLogStream;
+      executorService = Executors.newSingleThreadExecutor();
+    }
 
     @Override
     public void onRecovered(final ReadonlyStreamProcessorContext context) {
-      final int partitionId = context.getPartitionId();
-      typedEvent = new TypedRecordImpl(partitionId);
-      final var scheduleService = context.getScheduleService();
-
-      final LogStream logStream = context.getLogStream();
-      logStream.registerRecordAvailableListener(
-          () -> scheduleService.runDelayed(Duration.ZERO, this::onNewEventCommitted));
-      logStream
-          .newLogStreamReader()
-          .onComplete(
-              ((reader, throwable) -> {
-                if (throwable == null) {
-                  logStreamReader = reader;
-                  onNewEventCommitted();
-                }
-              }));
+      executorService.submit(
+          () -> {
+            final int partitionId = context.getPartitionId();
+            typedEvent = new TypedRecordImpl(partitionId);
+            final var asyncLogStream = synchronousLogStream.getAsyncLogStream();
+            asyncLogStream.registerRecordAvailableListener(this::onNewEventCommitted);
+            logStreamReader = asyncLogStream.newLogStreamReader().join();
+            exportEvents();
+          });
+    }
+
+    @Override
+    public void onClose() {
+      executorService.shutdownNow();
     }
 
     private void onNewEventCommitted() {
+      // this is called from outside (LogStream), so we need to enqueue the task
+      if (executorService.isShutdown()) {
+        return;
+      }
+
+      executorService.submit(this::exportEvents);
+    }
+
+    private void exportEvents() {
+      // to always run in the same thread
```

Comment:
:wrench: looks like the comment is a left-over.

Reply:
Ah okay, I thought you experimented with submitting to the executor here and that's where this comment got introduced. Then maybe change the comment to "skip if not yet recovered" or something along those lines?